text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,165 @@
+   intermediaries, according to intermediary instrumentations. Intermediaries
+   that are not instrumented might simply drop the transport context.
+
+   This layer helps to gain insights into details of the message transport.
+
+A producer MUST attach a creation context to each message. The creation context
+MUST be attached in a way so that it is not changed by intermediaries. A
+producer MAY propagate a transport context to an intermediary.  An
+intermediary MAY propagate a transport context to a consumer.
+
+### Span structure, names, and attributes
+
+#### Consumer
+
+For many use cases, it is not possible to rely on the presence of "Process"
+spans for correlating producer with consumer traces: there are cases where a
+dedicated processing operation cannot be identified, or where processing
+happens in a different trace. Furthermore, processing operations often are not
+covered by messaging libraries and SDKs, but take place in application code.
+Consistently creating spans for "Processing" operations would require either
+effort from the application owner to correctly instrument those operations, or
+additional capabilities of messaging libraries and SDKs (e. g. hooks for
+processing callbacks, which can then be instrumented by the libraries or SDKs).
+
+While it is possible to create "Process" spans and correlate those with
+consumer traces in certain cases, this is not something that can be generally
+required. Therefore, it is more feasible to require the creation of "Receive"
+spans to correlate producer with consumer traces. A "Receive" span must link
+to the "Create" spans of all messages that are handled by the respective
+"Receive" operation. Depending on the use case, "Receive" spans can correlate
+with "Process" spans or other spans modelling processing operations.
+
+The operation modelled by the "Receive" span does not strictly refer to
+receiving the message from intermediaries, but instead refers to the
+application receiving messages for processing. If messages are fetched from the
+intermediary and forwarded to the application in one go, the whole operation
+might be covered by a "Receive" span. However, clients might pre-fetch messages
+from intermediaries and cache those messages, and only forward messages to the
+application at a later time. In this case, the operation of pre-fetching and
+caching should not be covered by the "Receive" span.
+
+"Receive" spans SHOULD be created for all messages obtained by or passed to the
+application for processing. "Receive" spans MUST NOT be created for messages
+not forwarded to the application, but pre-fetched or cached by messaging
+libraries or SDKs. A single "Receive" span can account for a single message,
+for multiple messages (in case messages are passed for processing as batches),
+or for no message at all (in it is signalled that no messages were received).
+For each message it accounts for, a "Receive" span SHOULD link to the "Create"
+span for the message.
```

Comment:
The receive span works very well with batches, as it groups the entire messages batch under a single scope.
How do we imagine the trace for (very common) SDKs which expose an api for handling a single message each time?
Should there be a dummy "receive" span with exactly one message and no "process" spans?
Can we distinguish the "batch of one message" case from the "api for single message processing" case?

[Example from kafkajs](https://kafka.js.org/docs/consuming#a-name-each-message-a-eachmessage) - kafka client for nodejs:

### Single Message:
```js
await consumer.run({
    eachMessage: async ({ topic, partition, message }) => {
        console.log({
            key: message.key.toString(),
            value: message.value.toString(),
            headers: message.headers,
        })
    },
})
```

### Batch
```js
await consumer.run({
    eachBatchAutoResolve: true,
    eachBatch: async ({
        batch,
        resolveOffset,
        heartbeat,
        commitOffsetsIfNecessary,
        uncommittedOffsets,
        isRunning,
        isStale,
    }) => {
        for (let message of batch.messages) {
            console.log({
                topic: batch.topic,
                partition: batch.partition,
                highWatermark: batch.highWatermark,
                message: {
                    offset: message.offset,
                    key: message.key.toString(),
                    value: message.value.toString(),
                    headers: message.headers,
                }
            })

            resolveOffset(message.offset)
            await heartbeat()
        }
    },
})
```

Reply:
> How do we imagine the trace for (very common) SDKs which expose an api for handling a single message each time?
Should there be a dummy "receive" span with exactly one message and no "process" spans?

Good questions. Referencing the examples above, the current draft could be interpreted as follows:
* A "Receive" spans is created for each invocation of `eachBatch` or `eachMessage` callbacks. The duration is the duration of the callback invocation.
* It's the responsibility of the callback implementor to create "Process" spans. Those could either be children of the "Receive" span in the case of synchronous processing, or they could be links in the case of asynchronous processing.

> Can we distinguish the "batch of one message" case from the "api for single message processing" case?

If we want to be able to make this distinction, it is probably best to do this via attributes on "Receive" spans (e. g. `batch_size`).

Reply:
But having an attribute like `batch_size` doesn't really answer if it is in fact a batch right? If we use the attribute, and the first example above it will be `batch_size=1`. Maybe we can suggest adding the attribute only when implementations are really batching?

Reply:
> Can we distinguish the "batch of one message" case from the "api for single message processing" case?

I think this is the main open question in this thread. I think it might be beneficial to handle this via attributes (and not span structure or span names). However, this will be a point to discuss in the workgroup meetings.